File: app/src/main/java/com/travlog/android/apps/ui/activities/LinkedAccountsActivity.java
package com.travlog.android.apps.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.util.Pair;
import android.view.View;

import com.facebook.login.LoginManager;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.GoogleApiClient;
import com.jakewharton.rxbinding2.view.RxView;
import com.travlog.android.apps.R;
import com.travlog.android.apps.libs.BaseActivity;
import com.travlog.android.apps.libs.qualifiers.RequiresActivityViewModel;
import com.travlog.android.apps.viewmodels.LinkedAccountsViewModel;

import java.util.Arrays;

import butterknife.BindView;
import timber.log.Timber;

import static com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_IN_WITH_GOOGLE;
import static com.travlog.android.apps.libs.utils.TransitionUtilsKt.slideInFromLeft;

@RequiresActivityViewModel(LinkedAccountsViewModel.class)
public class LinkedAccountsActivity extends BaseActivity<LinkedAccountsViewModel> {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.facebook)
    View facebookButton;
    @BindView(R.id.google)
    View googleButton;

    private GoogleApiClient googleApiClient;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.a_linked_accounts);
        setSupportActionBar(toolbar);
        setDisplayHomeAsUpEnabled(true);

        final GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();

        googleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this,
                        connectionResult -> Timber.d("onConnectionFailed: %s", connectionResult.getErrorMessage()))
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();

        RxView.clicks(facebookButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> LoginManager.getInstance().logInWithReadPermissions(this, Arrays.asList("email", "public_profile")));

        RxView.clicks(googleButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> this.startSignInWithGoogleActivity());
    }

    private void startSignInWithGoogleActivity() {
        final Intent intent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient);
        startActivityForResult(intent, SIGN_IN_WITH_GOOGLE);
    }

    @Nullable
    @Override
    protected Pair<Integer, Integer> exitTransition() {
        return slideInFromLeft();
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/activities/LinkedAccountsActivity.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.ui.activities

import android.os.Bundle
import android.util.Pair
import com.facebook.login.LoginManager
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.GoogleApiClient
import com.jakewharton.rxbinding3.view.clicks
import com.travlog.android.apps.R
import com.travlog.android.apps.ViewModelFactory
import com.travlog.android.apps.getAppInjector
import com.travlog.android.apps.getViewModel
import com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_IN_WITH_GOOGLE
import com.travlog.android.apps.libs.BaseActivity
import com.travlog.android.apps.libs.utils.TransitionUtils.slideInFromLeft
import com.travlog.android.apps.viewmodels.LinkedAccountsViewModel
import kotlinx.android.synthetic.main.a_linked_accounts.*
import timber.log.Timber
import java.util.*
import javax.inject.Inject

class LinkedAccountsActivity : BaseActivity<LinkedAccountsViewModel>() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    private var googleApiClient: GoogleApiClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        getAppInjector().inject(this)
        viewModel = getViewModel(viewModelFactory)

        super.onCreate(savedInstanceState)

        setContentView(R.layout.a_linked_accounts)
        setSupportActionBar(toolbar)
        setDisplayHomeAsUpEnabled(true)

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build()

        googleApiClient = GoogleApiClient.Builder(this)
                .enableAutoManage(this) { connectionResult -> Timber.d("onConnectionFailed: %s", connectionResult.errorMessage) }
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build()

        facebook.clicks()
                .compose(bindToLifecycle())
                .subscribe { LoginManager.getInstance().logInWithReadPermissions(this, Arrays.asList("email", "public_profile")) }

        google.clicks()
                .compose(bindToLifecycle())
                .subscribe { this.startSignInWithGoogleActivity() }
    }

    private fun startSignInWithGoogleActivity() {
        val intent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient)
        startActivityForResult(intent, SIGN_IN_WITH_GOOGLE)
    }

    override fun exitTransition(): Pair<Int, Int>? {
        return slideInFromLeft()
    }
}
